src/Sarif.ValidationTests/DefaultObjectPopulatingVisitor.cs

```diff
@@ -0,0 +1,293 @@
+    /// whatever level in the tree is visited. This has the effect of comprehehsively 
+    /// 'hydrating' all possible SARIF constructs as defined by the C# OM.
+    /// </summary>
+    public class DefaultObjectPopulatingVisitor : SarifRewritingVisitor
+    {
+        public delegate object PrimitiveValueBuilder(bool isRequired);
+
+        // The default builders create a placeholder value for all required property primitives
+        // and emits default (null, 0, false, etc.) values for the remainder. As a result, a
+        // default hydration of this class should contain no non-default values for 
+        // JSON primitive-valued properties that aren't required
+        public static IDictionary<Type, PrimitiveValueBuilder> PopulateRequiredPrimitives()
+        {
+            return new Dictionary<Type, PrimitiveValueBuilder>()
+            {
+                { typeof(bool),    (isRequired) => { return isRequired; } },
+                { typeof(int),     (isRequired) => { return isRequired ? int.MaxValue : 0; } },
+                { typeof(double),  (isRequired) => { return isRequired ? double.MaxValue : 0; } },
+                { typeof(string),  (isRequired) => { return isRequired ? "Required string value." : null; } },
+                { typeof(DateTime),(isRequired) => { return isRequired ? DateTime.UtcNow : new DateTime(); } },
+                { typeof(Uri),     (isRequired) => { return isRequired ? new Uri("https://required.uri.contoso.com") : null; } }
+            };
+        }
+
+        // These builders exhaustively populate the generated SARIF file, including non-required JSON primitives
+        public static IDictionary<Type, PrimitiveValueBuilder> PopulateAllPrimitives()
+        {
+            // The values in this output are designed to allow conformance to the schema. In our schema,.
+            // for example, some integer values must be > 0. And so, out optional integer value is 
+            // written as 1.
+            return new Dictionary<Type, PrimitiveValueBuilder>()
+            {
+                { typeof(bool),    (isRequired) => { return true; } },
+                { typeof(int),     (isRequired) => { return isRequired ? int.MaxValue : 1; } },
+                { typeof(double),  (isRequired) => { return isRequired ? double.MaxValue : 1; } },
+                { typeof(string),  (isRequired) => { return isRequired ? "[Required string value]" : "[Optional string value]"; } },
+                { typeof(DateTime),(isRequired) => { return isRequired ? DateTime.UtcNow : new DateTime(1776, 7, 4).ToUniversalTime(); } },
+                { typeof(Uri),     (isRequired) => { return isRequired ? new Uri("https://required.uri.value.contoso.com") : new Uri("https://optional.uri.value.contoso.com"); } }
+            };
+        }
+
+        public DefaultObjectPopulatingVisitor(JsonSchema schema, IDictionary<Type, PrimitiveValueBuilder> primitiveValueBuilders = null)
```

Comment:
>primitiveValueBuilders [](start = 106, length = 22)

I see, so you can call this by passing the return value of one of the two above methods.

I don't think that's easy to understand. It unnecessarily exposes those two methods. For that matter, it unnecessarily exposes the delegate time `PrimitiveValueBuilder`. Consider instead using creation methods with revealing names. Make the ctor private, make the builder-dictionary-creating methods private, and make the delegate private. Then define:

```
public static DefaultObjectPopulatingVisitor CreateVisitorToPopulateRequiredPrimitivesOnly(JsonSchema schema);
public static DefaultObjectPopulatingVisitor CreateVisitorToPopulateAllPrimitives(JsonSchema schema);
```
... and the creation methods invoke the ctor with the appropriate set of builders.

Reply:
Will fix creation pattern in future change.

---
In reply to: [229108615](https://github.com/Microsoft/sarif-sdk/pull/1121#discussion_r229108615) [](ancestors = 229108615)

Reply:
Yes, what you say is why the design is still open. We need to author a few more tests (or at least I do) before all the requirements for customizing the default value emit sink in. It appears most tests will start with some baseline builders and go from there. It is very likely the delegate will change, for example, it likely needs to accept a JsonSchema object, not just the isRequired bool (but YAGNI dictates I start with the more focused argument). I suspect the 'real' change to make is to stop picking up default args with the populating visitor, which will force everyone to go through/discover the default builders.

---
In reply to: [229114719](https://github.com/Microsoft/sarif-sdk/pull/1121#discussion_r229114719) [](ancestors = 229114719,229108615)